File: neetcode150/easy/572-Subtree-of-Another-Tree.py
#!/usr/bin/python3
# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right
class Solution:
  # Initial Solution
  def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
    if not p and not q:
      return True
    elif not p or not q or p.val != q.val:
      return False
    else:
      return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
  def isSubtree(self, root: TreeNode, subRoot: TreeNode) -> bool:
    if not root and not subRoot:
      return True
    if not root or not subRoot:
      return False
    else:
      return (self.isSameTree(root, subRoot) or 
              self.isSubtree(root.left, subRoot) or 
              self.isSubtree(root.right, subRoot))

File: neetcode150/easy/572-Subtree-of-Another-Tree.cpp
#include <algorithm>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isSameTree(TreeNode* t1, TreeNode* t2) {
        if (t1 == NULL && t2 == NULL) {
            return true;
        } else if (t1 == NULL || t2 == NULL || t1->val != t2->val) {
            return false;
        } else {
            return isSameTree(t1->left, t2->left) && isSameTree(t1->right, t2->right);
        }
    }
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (root == NULL && subRoot == NULL) {
            return true;
        }
        if (root == NULL || subRoot == NULL) {
            return false;
        }
        return isSubtree(root->right, subRoot) || isSubtree(root->left, subRoot) || isSameTree(root, subRoot);
    }
};